### PR TITLE
`Programming exercises`: Fix an issue with git-diff report generation if it does not exist already

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Do one of these:
    - Create a programming exercise (and maybe wait for the build to finish)
    - Import a programming exercise (and maybe wait for the build to finish)
    - Locally: Delete the git-diff report of a programming exercise in the database
4. Go to the exercise detail page
5. The git-diff report should be (re)generated and shown at the bottom

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
Unchanged

### Screenshots

The git-diff stats that should be visible:
grafik

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java [hunk 1]
```diff
@@ -141,6 +141,23 @@ else if (reports.size() == 1) {
...
+     * @return The report or null if none can be generated
+     */
+    public ProgrammingExerciseGitDiffReport getOrCreateReportOfExercise(ProgrammingExercise programmingExercise) {
+        var report = this.getReportOfExercise(programmingExercise);
+        if (report == null) {
+            return updateReport(programmingExercise);
+        }
+        else {
+            return report;
+        }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:158]:
```suggestion
        return Objects.requireNonNullElseGet(report, () -> updateReport(programmingExercise));
```

Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:158]: I've looked at `requireNonNullElseGet` and it won't work here, as `updateReport` can return null and `requireNonNullElseGet` would throw a NullPointerException in these cases.
`updateReport` returns null when the exercise has not been fully setup yet (e.g. there is no Submission for the template/solution repos yet)

**Review by Strohgelaender — DISMISSED** [review]
code looks good!

**Review by ge65cer — COMMENTED** [review]
Tested on TS1. After creating a new programming exercise I waited about two minutes and the diff button did not appear. Only after editing the exercise and saving the changes did it show up. When importing the same exercise, it immediately appeared. Intended behavior?

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java (hunk 1 above)
Comment (JohannesStoehr) [reply src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java:158]: Some other code Parts use Optional.ofNullable(...).orElse(...)
Like
```
public Participant getParticipant() {
    return Optional.ofNullable((Participant) student).orElse(team);
}
```

**Review by JohannesStoehr — COMMENTED** [review]
Should we make getReportOfExercise private and force other classes to use getOrCreateReportOfExercise? Or is there any reason for keeping it public?

**Review by JohannesStoehr — APPROVED** [review]
Code changes look good to me

**Review by b-fein — APPROVED** [review]
Code changes look good.

**Review by krusche — APPROVED** [review]

**Pull request merged**